Revise: select profiles by typing the name, keep profile icons, new colors

SlySven opened the pull request:
This is so that the keyboard selection of profiles beginning with a particular letter works again. This will close issue #4283 .

This also returns it to a readable size instead of trying to hide it by setting the font size to 1 - as that would still show something (and a font size of zero is NOT valid). Manipulating the QPalette of the `QListWidget` does not work well enough in the presence of Styles or StyleSheets to be a viable means of hiding the text underneath either.

Also in the `dlgConnection` class:
* refactor out some highly repetitive code that sets up the predefined Mud icons.
* remove warnings and adopts a change introduced into Qt 5.15 for the `QLabel::pixmap()` which now requires a `Qt::ReturnByValue` argument to prevent an obsolete code warning and to force a `QPixmap` to be returned by value and not reference...
* remove a warning about: `QListWidget::setItemSelected(QListWidgetItem*,bool)` being deprecated and switches to the recommended: `QListWidgetItem::setSelected(bool)`

Take the display of the name from the custom icon - as we now show it visibly underneath and if the user assigns their own icon the text is not shown anyway.

Change the custom icon generator so that there is NOT a dependency on the position in the list for the colours used. This means that the same name will always have the same colours - which will help the user to pick out a particular profile as it will always look the same no matter if other profiles are added or removed.

Change the colour gradient to be a radial one rather than a linear one, this makes it more *interesting* I think.

Refactor out the colour icon generator to a single method that is called from more than one place.

Closes https://github.com/Mudlet/Mudlet/issues/4288

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: restore profile name to connection icon

This is so that the keyboard selection of profiles beginning with a
particular letter works again. This will close issue #4283 .

This also returns it to a readable size instead of trying to hide it by
setting the font size to 1 - as that would still show something (and a font
size of zero is NOT valid). Manipulating the QPalette of the `QListWidget`
does not work well enough in the presence of Styles or StyleSheets to be
a viable means of hiding the text underneath either.

Also in the `dlgConnection` class:
* refactor out some highly repetitive code that sets up the predefined
Mud icons.
* remove warnings and adopts a change introduced into Qt 5.15 for
the `QLabel::pixmap()` which now requires a `Qt::ReturnByValue` argument
to prevent an obsolete code warning and to force a QPixmap to be returned
by value and not reference...
* remove a warning about:
  `QListWidget::setItemSelected(QListWidgetItem*,bool)` being deprecated
  and switches to the recommended:
  `QListWidgetItem::setSelected(bool)`

Take the display of the name from the custom icon - as we now show it
visibly underneath and if the user assigns their own icon the text is not
shown anyway.

Change the custom icon generator so that there is NOT a dependency on the
position in the list for the colours used. This means that the same name
will always have the same colours - which will help the user to pick out
a particular profile as it will always look the same no matter if other
profiles are added or removed.

Change the colour gradient to be a radial one rather than a linear one,
this makes it more *interesting* I think.

Refactor out the colour icon generator to a single method that is called
from more than one place.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: keep CodeFactor happy...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: restore odd original way of accessing some QPixmaps

This was causing failures in some CI platforms using Qt 5.14.1.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
The community isn't sold on expanding the text. Let's keep it as before. I understand it's a problem for you, but nobody else mentioned it before being an issue, so it seems like a relatively minor one in the grand scheme of themes. People have on the other hand mentioned the inability to type to select profiles though.

SlySven committed:
Revise: accept peer request - put text back on icons & shrink it under them

This should almost restore the original arrangement (though I am not going
to bother to try setting the text explicitly to white as that just doesn't
work for multiple reasons).

This still means that there are some (coding) enhancements in place - as
well as the slightly funky gradients on custom icons.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
I personally like the gradient refresh, and the fact that it won't change now if you add/remove profiles :+1: 

I think sometimes it picks colors that are too similar and thus it generates an icon that looks rather plain compared to others:

![Selection_510](https://user-images.githubusercontent.com/110988/101487550-a7cef780-395e-11eb-877d-e575964fd9f4.png)
![Selection_511](https://user-images.githubusercontent.com/110988/101487561-aa315180-395e-11eb-8778-e195856f8379.png)
![Selection_512](https://user-images.githubusercontent.com/110988/101487581-aef60580-395e-11eb-9cec-1a53630cb6a0.png)
![Selection_513](https://user-images.githubusercontent.com/110988/101487594-b2898c80-395e-11eb-9c04-d2ea7b563db2.png)

SlySven commented:
> I think sometimes it picks colors that are too similar and thus it generates an icon that looks rather plain compared to others:

Well it is too random, it is just a 32 (or possibly 64 bit) hash of the profile name that is multiplied now by 6 different prime numbers and then each of them, modulus 256 is used to specify one of the colour components at each of the two ends of the `QRadiusGradient` - I suppose I *could* try another three to specify an additional point somewhere on that gradient but I think you will just be on a journey of diminishing returns...

vadi2 commented:
> is just a 32 (or possibly 64 bit) hash of the profile name

Can you make it so it's the same color no matter what computer is used?

SlySven commented:
> Can you make it so it's the same color no matter what computer is used?

I think it will be as Qt `qHash(...)` methods are used - see **[`(uint) qHash(const QString &key, uint seed=0)`](https://doc.qt.io/qt-5/qhash.html#qHash-17)**.

SlySven committed:
Merge branch 'development' into Revise_restoreProfileNameToConnectionsIcon

SlySven committed:
Revise: really frob the custom icon generator

This is to try and produce as varied but still readable icons when they are
generated entirely on the basis of the MUD name - in a reproducible
(though Qt do NOT guarantee repeatable qHash() functionality across Qt
versions - so they may not say *that* constant) manner.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Refrob: changes things back a bit but leave outline on text on custom icons

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -2316,14 +2308,25 @@ QIcon dlgConnectionProfiles::customIcon(const QString& text) const
         pt.drawText(QRect(30, 0, 89, 29), Qt::AlignCenter|Qt::TextSingleLine, text);
-        // Then draw the text in white centered on top of the black ones:
-        pt.setPen(Qt::white);
+        // Then draw the text in the opposite used before to get a strong
```
```suggestion
        // Then draw the text in the opposite color than used before to get a strong
```

SlySven committed:
Update src/dlgConnectionProfiles.cpp

Accept a peer suggestion.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

Co-authored-by: Kebap <kebap_spam@gmx.net>

SlySven committed:
Refrob: use a table of fixed colours

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Final: use single but less fixed hues and with three saturations

Using only 15 hues but with 3 different saturation level plus 5 grey-scale
one should give enough different icons but not generate them too close to
each other for those with normal colour vision...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Cleanup: remove a bit of cruft I left behind whilst hacking this PR.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (approved):
Excellent work!

vadi2 commented:
Design things seem to take on quite a few iterations, thanks for having the patience for it :)

Kebap commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -271,6 +283,33 @@ dlgConnectionProfiles::dlgConnectionProfiles(QWidget * parent)
+
+    // Add some color ones with evenly spaced hue
+    for (quint16 i = 0; i < 360; i += 24) {
+        for (quint16 j = 0; j < 3; ++j) {
+            switch (j) {
+            case 0: {
+                mCustomIconColors.append(QColor::fromHsv(i, 255, 255));
+                break;
+            }
+            case 1: {
+                mCustomIconColors.append(QColor::fromHsv(i, 192, 255));
+                break;
+            }
+            case 2: {
+                mCustomIconColors.append(QColor::fromHsv(i, 128, 255));
+                break;
+            }
+            }
```
```suggestion
        mCustomIconColors.append(QColor::fromHsv(i, 255, 255));
        mCustomIconColors.append(QColor::fromHsv(i, 192, 255));
        mCustomIconColors.append(QColor::fromHsv(i, 128, 255));
```
